### PR TITLE
Support pytorch GPTQ: STE algorithm

### PR DESCRIPTION
**Pull request by eladc-git**
Phase1: Only STE, no support for Gumbel rounding yet.

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py [hunk 1]
```diff
@@ -125,7 +141,8 @@ class PytorchModel(torch.nn.Module):
     def __init__(self,
                  graph: Graph,
                  append2output: List[Any] = None,
-                 fw_info: FrameworkInfo = DEFAULT_PYTORCH_INFO):
+                 fw_info: FrameworkInfo = DEFAULT_PYTORCH_INFO,
```
Comment (haihabi) [inline model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:144]: Use default framework info as an import and not input

File: tests/common_tests/helpers/tensors_compare.py [hunk 2]
```diff
@@ -58,3 +58,7 @@ def tensor_compare(a, b):
     cs_value = cosine_similarity(a, b)
     norm_rate = tensor_norm(a) / tensor_norm(b)
     return cs_value, norm_rate
+
+
+def calc_db(x, eps=1e-10):
```
Comment (reuvenperetz) [inline tests/common_tests/helpers/tensors_compare.py:63]: Comments and type hints.

File: tests/pytorch_tests/model_tests/feature_models/gptq_test.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    def run_test(self):
+        # Create model
+        self.float_model = self.create_networks()
+
+        # Create quantization config
+        qConfig = self.get_quantization_config()
+
+        # Run MCT with PTQ
+        np.random.seed(self.seed)
+        ptq_model, ptq_qInfo = mct.pytorch_post_training_quantization_experimental(self.float_model,
```
Comment (reuvenperetz) [inline tests/pytorch_tests/model_tests/feature_models/gptq_test.py:64]: Please rename ptq_qInfo to follow conventions (maybe ptq_user_info or something like that)

File: tests/pytorch_tests/model_tests/feature_models/gptq_test.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            else mct.pytorch_post_training_quantization(self.float_model,
+                                                        self.representative_data_gen,
+                                                        n_iter=self.num_calibration_iter,
+                                                        quant_config=qConfig,
+                                                        fw_info=DEFAULT_PYTORCH_INFO,
+                                                        network_editor=self.get_network_editor())
+
+        # Run MCT with GPTQ
+        np.random.seed(self.seed)
+        gptq_model, gptq_qInfo = mct.pytorch_gradient_post_training_quantization_experimental(self.float_model,
```
Comment (reuvenperetz) [inline tests/pytorch_tests/model_tests/feature_models/gptq_test.py:78]: Please rename variable

File: model_compression_toolkit/gptq/pytorch/gptq_training.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            self.input_scale = self.gptq_user_info.input_scale
+
+        trainable_weights, trainable_bias, trainable_threshold = get_trainable_parameters(self.fxp_model,
+                                                                                          add_bias=self.gptq_config.train_bias,
+                                                                                          quantization_parameters_learning=self.gptq_config.quantization_parameters_learning)
+        self.flp_weights_list, self.fxp_weights_list = get_weights_for_loss(self.fxp_model)
+        if not (len(self.compare_points) == len(trainable_weights) == len(self.flp_weights_list) == len(
+                self.fxp_weights_list)):
+            raise Exception(
+                "GPTQ: Mismatch between number of compare points, number of layers with trainable weights " +
```
Comment (reuvenperetz) [inline model_compression_toolkit/gptq/pytorch/gptq_training.py:72]: Please use the logger instead of raising an exception.

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py [hunk 6]
```diff
@@ -76,24 +75,41 @@ def _run_operation(n: BaseNode,
...
-        out_tensors_of_n = op_func(*input_tensors + op_call_args, **functional_kwargs)
+        out_tensors_of_n_float = op_func(*input_tensors + op_call_args, **functional_kwargs)
 
     # Add a fake quant node if the node has an activation threshold.
+    out_tensors_of_n = out_tensors_of_n_float
     if n.is_activation_quantization_enabled():
-        out_tensors_of_n = quantize_node_activation_fn(n, out_tensors_of_n)
+        if isinstance(out_tensors_of_n_float, list):
+            out_tensors_of_n_float = torch.cat(out_tensors_of_n_float, dim=0)
+        out_tensors_of_n = quantize_node_activation_fn(n, out_tensors_of_n_float)
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:96]: because we need it for out_tensors_of_n_float also

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py [hunk 7]
```diff
@@ -66,7 +65,7 @@ def _build_input_tensors_list(node: BaseNode,
 def _run_operation(n: BaseNode,
                    input_tensors: List,
                    op_func: Any,
-                   quantize_node_activation_fn) -> List[torch.Tensor]:
+                   quantize_node_activation_fn) -> Tuple[Any, Any]:
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:68]: because the output: (out_tensors_of_n, out_tensors_of_n_float) can be torch.Tensor or list of torch.Tensor

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py [hunk 8]
```diff
@@ -76,24 +75,41 @@ def _run_operation(n: BaseNode,
...
-        out_tensors_of_n = quantize_node_activation_fn(n, out_tensors_of_n)
+        if isinstance(out_tensors_of_n_float, list):
+            out_tensors_of_n_float = torch.cat(out_tensors_of_n_float, dim=0)
+        out_tensors_of_n = quantize_node_activation_fn(n, out_tensors_of_n_float)
 
-    return out_tensors_of_n
+    return out_tensors_of_n, out_tensors_of_n_float
+
+
+def _find_by_node_name(node_to_output_tensors_dict: dict, node_name: str):
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:101]: I wish. We need to find node by name and not by node type.

File: model_compression_toolkit/core/pytorch/utils.py [hunk 9]
```diff
@@ -68,7 +72,7 @@ def to_torch_tensor(tensor):
         raise Exception(f'Conversion of type {type(tensor)} to {type(torch.Tensor)} is not supported')
 
 
-def torch_tensor_to_numpy(tensor: torch.Tensor) -> np.ndarray:
+def torch_tensor_to_numpy(tensor: [torch.Tensor, list, tuple]) -> [np.ndarray, list, tuple]:
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/core/pytorch/utils.py:75]: I think Union is more relevant

File: model_compression_toolkit/gptq/pytorch/quantizer/gptq_quantizer.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        if weights_quantization_cfg.weights_quantization_method == QuantizationMethod.UNIFORM:
+            # Uniform quantization
+            self.min_range = weights_quantization_cfg.weights_quantization_params.get(RANGE_MIN)
+            self.max_range = weights_quantization_cfg.weights_quantization_params.get(RANGE_MAX)
+        else:
+            # Symmetric quantization
+            threshold_values = weights_quantization_cfg.weights_quantization_params.get(THRESHOLD)
+            self.min_range = -threshold_values
+            self.max_range = threshold_values
+        self.delta_tensor = (self.max_range - self.min_range) / (2 ** self.num_bits)
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/gptq/pytorch/quantizer/gptq_quantizer.py:53]: Right. I am removing Uniform quantization. Only PO2 and sym.

File: model_compression_toolkit/gptq/pytorch/quantizer/quant_utils.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    """
+    Clip a variable between fixed values such that min_val<=output<=max_val
+    Args:
+        x: input variable
+        min_val: minimum value for clipping
+        max_val: maximum value for clipping
+    Returns:
+        clipped variable
+    """
+    return (torch.clip(x, min=min_val, max=max_val) - x).detach() + x
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/gptq/pytorch/quantizer/quant_utils.py:38]: This is the pytorch implementation of STE: detach() is ignore in forward pass, and zero in backward pass.

File: model_compression_toolkit/gptq/pytorch/gptq_graph_info.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+            trainable_aux_weights.append(layer.weight_quantizer.get_aux_variable())
+            if quantization_parameters_learning:
+                trainable_threshold.append(layer.weight_quantizer.get_quantization_variable())
+            if add_bias and layer.op.bias is not None:
+                trainable_bias.append(layer.op.bias)
+
+    return trainable_aux_weights, trainable_bias, trainable_threshold
+
+
+def get_weights_for_loss(fxp_model: nn.Module) -> ([List[list], List[list]]):
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/gptq/pytorch/gptq_graph_info.py:49]: fixed!

File: tests/common_tests/helpers/tensors_compare.py (hunk 2 above)
Comment (eladc-git) [reply tests/common_tests/helpers/tensors_compare.py:63]: Since when we are using "Comments and type hints" in tests?

File: model_compression_toolkit/gptq/pytorch/quantization_facade.py [hunk 13]
```diff
@@ -13,25 +13,84 @@
...
+from model_compression_toolkit.core.exporter import export_model
+from model_compression_toolkit.core.analyzer import analyzer_model_quantization
 from model_compression_toolkit import CoreConfig
+from model_compression_toolkit.core.common.mixed_precision.mixed_precision_quantization_config import \
+    MixedPrecisionQuantizationConfigV2
+
+import importlib
 
-if FOUND_TORCH:
+if importlib.util.find_spec("torch") is not None:
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/gptq/pytorch/quantization_facade.py:34]: I will put it back. I have some issue in github checks: "torch no found"

**Review by haihabi — APPROVED** [review]

File: model_compression_toolkit/gptq/pytorch/quantizer/gptq_quantizer.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+from abc import abstractmethod
+
+class BaseWeightQuantizer(nn.Module):
+
+    def __init__(self):
+        """
+        Construct a Base Pytorch model that utilizes a fake weight quantizer
+        """
+        super().__init__()
+        self.trainable_params = dict()
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/gptq/pytorch/quantizer/gptq_quantizer.py:27]: I use them in QuantizerWrapper class

**Review by reuvenperetz — APPROVED** [review]
